homeassistant/components/aquacell/sensor.py

```diff
@@ -0,0 +1,157 @@
+
+from collections.abc import Callable
+from dataclasses import dataclass
+from datetime import datetime
+from typing import cast
+
+from aioaquacell import Softener
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import PERCENTAGE, UnitOfTime
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+from .coordinator import AquacellCoordinator
+from .entity import AquacellEntity
+
+PARALLEL_UPDATES = 1
+
+
+@dataclass(frozen=True, kw_only=True)
+class SoftenerEntityDescription(SensorEntityDescription):
+    """Describes Softener sensor entity."""
+
+    value_fn: Callable[[Softener], str | int | float | None]
+
+
+SENSORS: tuple[SoftenerEntityDescription, ...] = (
+    SoftenerEntityDescription(
+        key="salt_leftpercentage",
+        translation_key="salt_leftpercentage",
+        state_class=SensorStateClass.MEASUREMENT,
+        icon="mdi:basket-fill",
+        native_unit_of_measurement=PERCENTAGE,
+        value_fn=lambda softener: cast(int, softener.salt.leftPercent),
+    ),
+    SoftenerEntityDescription(
+        key="salt_rightpercentage",
+        translation_key="salt_rightpercentage",
+        icon="mdi:basket-fill",
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=PERCENTAGE,
+        value_fn=lambda softener: cast(int, softener.salt.rightPercent),
+    ),
+    SoftenerEntityDescription(
+        key="salt_leftdays",
+        translation_key="salt_leftdays",
+        icon="mdi:calendar-today",
+        native_unit_of_measurement=UnitOfTime.DAYS,
+        value_fn=lambda softener: softener.salt.leftDays,
+    ),
+    SoftenerEntityDescription(
+        key="salt_rightdays",
+        translation_key="salt_rightdays",
+        icon="mdi:calendar-today",
+        native_unit_of_measurement=UnitOfTime.DAYS,
+        value_fn=lambda softener: softener.salt.rightDays,
+    ),
+    SoftenerEntityDescription(
+        key="fw_version",
+        translation_key="fw_version",
+        icon="mdi:chip",
+        value_fn=lambda softener: softener.fwVersion,
+    ),
+    SoftenerEntityDescription(
+        key="name",
+        translation_key="name",
+        value_fn=lambda softener: softener.name,
+    ),
+    SoftenerEntityDescription(
+        key="battery",
+        translation_key="battery",
+        device_class=SensorDeviceClass.BATTERY,
+        native_unit_of_measurement=PERCENTAGE,
+        value_fn=lambda softener: softener.battery,
+    ),
+    SoftenerEntityDescription(
+        key="last_update",
+        translation_key="last_update",
+        icon="mdi:update",
+        value_fn=lambda softener: datetime.fromtimestamp(
+            cast(float, softener.lastUpdate) / 1000
+        ).isoformat(),
+    ),
```

Comment:
This is already stored in HA per state, so we should not add this one either

Reply:
This is reported from the device, the time when it has sent an update to the cloud. So when the data is polled you can see if the data is recent or not.

Reply:
Yes, but we don't allow these kinds of sensors in HA as HA already keeps track of that

Reply:
We might have some confusion here, let me explain more thoroughly:
The device sends updates to the cloud, usually once a day. And HA polls the cloud 'cached' values, so it does not retrieve directly from the device. So for example when the salt level stays on 30% for two weeks, you might want to know if it actually measured it recently. HA wouldn't know about this since there's no state change.

Reply:
But you have statistics and can see that the last time the data changed was x days ago.

Reply:
Proposal for now: Omit the sensor, you can add this in a separate PR so you can state your case and it doesn't hold up this PR

Reply:
Ok will do that.